components/core/src/clp/ffi/ir_stream/Deserializer.cpp

```diff
@@ -0,0 +1,704 @@
+ * @return The corresponded schema tree node type on success.
+ * @return std::nullopt if the tag doesn't match to any defined schema tree node types.
+ */
+[[nodiscard]] auto schema_tree_node_tag_to_type(encoded_tag_t tag
+) -> std::optional<SchemaTreeNode::Type>;
+
+/**
+ * Deserializes the parent ID of a schema tree node.
+ * @param reader
+ * @param parent_id Returns the deserialized result.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_schema_tree_node_parent_id(
+        ReaderInterface& reader,
+        SchemaTreeNode::id_t& parent_id
+) -> IRErrorCode;
+
+/**
+ * Deserializes an integer value packet.
+ * @param reader
+ * @param tag
+ * @param val Returns the deserialized value.
+ */
+[[nodiscard]] auto
+deserialize_int_val(ReaderInterface& reader, encoded_tag_t tag, value_int_t& val) -> IRErrorCode;
+
+/**
+ * Deserializes a string packet.
+ * @param reader
+ * @param tag A tag indicating the number of bytes used for the encoded length.
+ * @param deserialized_str Returns the deserialized string.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_string(
+        ReaderInterface& reader,
+        encoded_tag_t tag,
+        std::string& deserialized_str
+) -> IRErrorCode;
+
+/**
+ * Deserializes all UTC offset packets until a non-UTC offset packet tag is read.
+ * @param reader
+ * @param tag Takes the current tag as input and returns the last tag read.
+ * @param utc_offset Returns the deserialized UTC offset.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_utc_offset_changes(
+        ReaderInterface& reader,
+        encoded_tag_t& tag,
+        UtcOffset& utc_offset
+) -> IRErrorCode;
+
+/**
+ * Deserializes all the schema tree node packets and inserts them into the schema tree until a non
+ * schema tree node tag is read.
+ * @param reader
+ * @param tag Takes the current tag as input and returns the last tag read.
+ * @param schema_tree Returns the schema tree with all new nodes inserted.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_schema_tree_nodes(
+        ReaderInterface& reader,
+        encoded_tag_t& tag,
+        SchemaTree& schema_tree
+) -> IRErrorCode;
+
+/**
+ * Deserializes the schema (a vector of node IDs) of a log event.
+ * @param reader
+ * @param tag Takes the current tag as input and returns the last tag read.
+ * @param schema Returns the deserialized schema.
+ * @return IRErrorCode::IRErrorCode_Success on success.
+ * @return IRErrorCode indicating relevant errors.
+ */
+[[nodiscard]] auto deserialize_schema(
+        ReaderInterface& reader,
+        encoded_tag_t& tag,
+        std::vector<SchemaTreeNode::id_t>& schema
```

Comment:
I think for someone new to this code, `schema` might be a confusing/overloaded term (and it's not something that we've documented well in `SchemaTree`). So perhaps we should create a type alias and add a docstring for it explaining that it's a collection of schema tree leaf node IDs that represents the schema of a KV-pair log event.